src/project_model/project_manifest.rs

```diff
@@ -1,71 +1,58 @@
 // copy from rust-analyzer
 
+use std::path::{Path, PathBuf};
 use std::{
     fs::{self, read_dir, ReadDir},
     io,
 };
-use std::path::{Path, PathBuf};
 
-use anyhow::{bail, format_err, Result};
+use anyhow::Result;
 use rustc_hash::FxHashSet;
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
-pub struct ManifestPath {
-    file: PathBuf,
-}
+use crate::logger::log;
 
-impl TryFrom<PathBuf> for ManifestPath {
-    type Error = PathBuf;
-
-    fn try_from(file: PathBuf) -> Result<Self, Self::Error> {
-        if file.parent().is_none() {
-            Err(file)
-        } else {
-            Ok(ManifestPath { file })
-        }
-    }
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Default)]
+pub struct ProjectManifest {
+    pub file: PathBuf,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
-pub struct ProjectManifest(ManifestPath);
-
 impl ProjectManifest {
-    pub fn from_manifest_file(path: PathBuf) -> Result<ProjectManifest> {
-        let path = ManifestPath::try_from(path)
-            .map_err(|path| format_err!("bad manifest path: {}", path.display()))?;
-        if path.file.file_name().unwrap_or_default() == "Cargo.toml" {
-            return Ok(ProjectManifest(path));
-        }
-        bail!(
-            "project root must point to Cargo.toml {}",
-            path.file.display()
-        );
-    }
-
-    // TODO check how it works when cargo.toml not only in the the main folder
     pub fn discover(path: &PathBuf) -> io::Result<Vec<ProjectManifest>> {
```

Comment:
tutaj totalnie potrzebujecie testów żeby upewnić się, że to działa tak jak powinno: możecie użyć https://docs.rs/tempfile/latest/tempfile/fn.tempdir.html żeby sobie zrobić pliki wewnątrz testów

Reply:
fajna libka, testy już dodane :)